src/cuda/poisson_fft.f90

```diff
@@ -0,0 +1,239 @@
+   end type cuda_poisson_fft_t
+
+   interface cuda_poisson_fft_t
+      module procedure init
+   end interface cuda_poisson_fft_t
+
+contains
+
+   function init(xdirps, ydirps, zdirps) result(poisson_fft)
+      implicit none
+
+      class(dirps_t), intent(in) :: xdirps, ydirps, zdirps
+
+      type(cuda_poisson_fft_t) :: poisson_fft
+
+      integer :: nx, ny, nz
+
+      integer :: ierrfft
+      integer(int_ptr_kind()) :: worksize
+
+      call poisson_fft%base_init(xdirps, ydirps, zdirps, SZ)
+
+      nx = poisson_fft%nx; ny = poisson_fft%ny; nz = poisson_fft%nz
+
+      allocate (poisson_fft%waves_dev(SZ, nx, (ny*(nz/2 + 1))/SZ))
+      poisson_fft%waves_dev = poisson_fft%waves
+
+      allocate (poisson_fft%ax_dev(nx), poisson_fft%bx_dev(nx))
+      allocate (poisson_fft%ay_dev(ny), poisson_fft%by_dev(ny))
+      allocate (poisson_fft%az_dev(nz), poisson_fft%bz_dev(nz))
+      poisson_fft%ax_dev = poisson_fft%ax; poisson_fft%bx_dev = poisson_fft%bx
+      poisson_fft%ay_dev = poisson_fft%ay; poisson_fft%by_dev = poisson_fft%by
+      poisson_fft%az_dev = poisson_fft%az; poisson_fft%bz_dev = poisson_fft%bz
+
+      allocate (poisson_fft%c_x_dev(nx*ny*(nz/2 + 1)))
+      allocate (poisson_fft%c_y_dev(nx*ny*(nz/2 + 1)))
+      allocate (poisson_fft%c_z_dev(nx*ny*(nz/2 + 1)))
+
+      ! We can't currently ask allocator to pass us an array with
+      ! exact shape we want, so we allocate an extra array here.
+      ! This will be removed when allocator is fixed.
+      allocate (poisson_fft%f_tmp(nz, SZ, nx*ny/SZ))
+
+      ! set cufft plans
+      ierrfft = cufftCreate(poisson_fft%planD2Zz)
+      ierrfft = cufftMakePlanMany(poisson_fft%planD2Zz, 1, nz, &
+                                  nz, 1, nz, nz/2+1, 1, nz/2+1, &
+                                  CUFFT_D2Z, nx*ny, worksize)
+      ierrfft = cufftSetWorkArea(poisson_fft%planD2Zz, poisson_fft%c_x_dev)
+
+      ierrfft = cufftCreate(poisson_fft%planZ2Dz)
+      ierrfft = cufftMakePlanMany(poisson_fft%planZ2Dz, 1, nz, &
+                                  nz/2+1, 1, nz/2+1, nz, 1, nz, &
+                                  CUFFT_Z2D, nx*ny, worksize)
+      ierrfft = cufftSetWorkArea(poisson_fft%planZ2Dz, poisson_fft%c_x_dev)
+
+      ierrfft = cufftCreate(poisson_fft%planZ2Zy)
+      ierrfft = cufftMakePlanMany(poisson_fft%planZ2Zy, 1, ny, &
+                                  ny, 1, ny, ny, 1, ny, &
+                                  CUFFT_Z2Z, nx*(nz/2 + 1), worksize)
+      ierrfft = cufftSetWorkArea(poisson_fft%planZ2Zy, poisson_fft%c_x_dev)
+
+      ierrfft = cufftCreate(poisson_fft%planZ2Zx)
+      ierrfft = cufftMakePlanMany(poisson_fft%planZ2Zx, 1, nx, &
+                                  nx, 1, nx, nx, 1, nx, &
+                                  CUFFT_Z2Z, ny*(nz/2 + 1), worksize)
+      ierrfft = cufftSetWorkArea(poisson_fft%planZ2Zx, poisson_fft%c_y_dev)
+
+   end function init
+
+   subroutine fft_forward_cuda(self, f)
```

Comment:
Would be good to have working tests for these before trying to implement cuFFTMP.

Reply:
@semi-h implement in this PR or create an issue?

Reply:
Actually, cuFFTMp provides 3D forward and inverse FFT transforms, and our functions here will only carry out the calls to cuFFTMp library so probably there is no value testing them.

Reply:
For the forward/backward FFTs sure, no need to test individually. But we're currently not unit testing the Poisson solver at all. Since we have a working implementation, I think it would be easiest to implement a correct test of this working solver, then apply the same test to cuFFTMP.

If you don't already have one kicking around, the analytical test case I usually use for Poisson solve is solving $\nabla^2 p = -\sin(\pi x) \sin(\pi y)$ and comparing to analytical solution $p_{analytical}(x,y) = \sin (\pi x) \sin(\pi y)/(2\pi^2)$.